pywbem_mock/_providerdispatcher.py

```diff
@@ -102,6 +102,82 @@ def _array_ness(is_array):
         """Return text for array-ness"""
         return "an array" if is_array else "a scalar"
 
+    def _validate_property(
+            self, prop_name, instance, creation_class, namespace):
+        """
+        Validate a property of an instance against its declaration in a class.
+        """
+
+        if prop_name not in creation_class.properties:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Property {0!A} in the instance does not exist in its "
+                        "creation class {1!A} in namespace {2!A} of the CIM "
+                        "repository",
+                        prop_name, instance.classname, namespace))
+
+        prop_inst = instance.properties[prop_name]
+        prop_cls = creation_class.properties[prop_name]
+
+        if prop_inst.type != prop_cls.type:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Property {0!A} in the instance has incorrect "
+                        "type={1!A}, but should have type={2!A} according to "
+                        "its creation class {3!A} in namespace {4!A} of the "
+                        "CIM repository",
+                        prop_name, prop_inst.type, prop_cls.type,
+                        instance.classname, namespace))
+
+        if prop_inst.is_array != prop_cls.is_array:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Property {0!A} in the instance has incorrect "
+                        "is_array={1!A}, but should have is_array={2!A} "
+                        "according to its creation class {3!A} in namespace "
+                        "{4!A} of the CIM repository",
+                        prop_name, prop_inst.is_array, prop_cls.is_array,
+                        instance.classname, namespace))
+
+        if isinstance(prop_inst.value, CIMInstance):
+            emb_classname_inst = prop_inst.value.classname
+            if 'EmbeddedInstance' in prop_cls.qualifiers:
+                emb_classname_cls = prop_cls.qualifiers['EmbeddedInstance']
+                # TODO: Expand test to allow inheritance
```

Comment:
This tests the new or modified instance that is an input parameter of the CreateInstance or ModifyInstance client operation, against the creation class that is in the repository.

The client call may have been made by the user or by the MOF compiler. In case it is made by the user, the MOF compiler is not even involved for the input instance.

The same is true for the creation class that is already in the repository.

Maybe the MOF compiler should check less if it is not even always involved, and leave that to the client calls of the mocker, which are always involved.

Also, the subclass test is not very expensive, IMO. If we walk the superclasses of a given class up to the root, tha is sufficient for what we need here, and that is much better compared to determining all subclasses of a given class. And embedded instances are pretty rare.

DISCUSSION

Reply:
I have put up an update to the PR that introduces a BaseProvider.is_subclass() method (including testcases) and that uses it for the EI test.

Please review.

Reply:
The compiler must check everything except the is_subclass since it is the tool that compiles the property value.  I was thinking backwards when It thought of get_subclasses.  Good catch and yes, is_subclass should have been one of our standard methods long ago.

Reply:
ok